### ISSUE TITLE
Derived `Hash` impl is duck-typed

### ISSUE THREAD

**Issue by apasel422** [rust-lang/rust#21160]
The following compiles successfully with `rustc 1.0.0-nightly (170c4399e 2015-01-14 00:41:55 +0000)`:
```rust
struct Bar;
impl Bar {
    fn hash<T>(&self, _: T) {}
}
#[derive(Hash)]
struct Foo(Bar);
```
The derived code should probably use UFCS.

**Comment by apasel422** [comment]
I'm working on a fix for this.

**Comment by japaric** [comment]
fix in #21161 

@apasel422 sorry, didn't see your message! I should've left a comment here

**Comment by apasel422** [comment]
@japaric That's OK! Glad it's fixed. We should make sure the other deriving implementations don't suffer from the same problem.
